src/ewah/bitset_encoder.zig

```diff
@@ -0,0 +1,234 @@
+const std = @import("std");
+const assert = std.debug.assert;
+const mem = std.mem;
+
+/// `encode`/`decode` compress the bitset using EWAH.
+///
+/// "Histogram-Aware Sorting for Enhanced Word-Aligned Compression in Bitmap Indexes":
+///
+/// > [EWAH] also uses only two types of words, where the first type is a 64-bit verbatim word.
+/// > The second type of word is a marker word: the first bit indicates which clean word will follow,
+/// > half the bits (32 bits) are used to store the number of clean words, and the rest of the bits
+/// > (31 bits) are used to store the number of dirty words following the clean words. EWAH bitmaps
+/// > begin with a marker word.
+///
+/// A 'marker' looks like:
+///
+///     [run_bit:u1][run_length:u31(LE)][literals_count:u32(LE)]
+///
+/// and is immediately followed by `literals_count` 64-bit literals.
+/// When decoding a marker, the run precedes the literals.
+///
+/// This encoding requires that the architecture is little-endian with 64-bit words.
+pub fn BitSetEncoder(comptime Word: type) type {
+    const word_bits = @bitSizeOf(Word);
+
+    const marker_run_length_max = (1 << (word_bits / 2)) - 1;
+    const marker_literals_max = (1 << (word_bits / 2 - 1)) - 1;
+
+    const MarkerRunLength = std.math.IntFittingRange(0, marker_run_length_max); // Word=usize → u32
+    const MarkerLiterals = std.math.IntFittingRange(0, marker_literals_max); // Word=usize → u31
+
+    const Marker = packed struct {
+        run_bit: u1, // run of 0s or 1s
+        literals: MarkerLiterals, // number of literal words immediately following marker
+        run_length: MarkerRunLength, // length of run (in words)
+    };
+
+    comptime {
+        assert(std.Target.current.cpu.arch.endian() == std.builtin.Endian.Little);
+        assert(@typeInfo(Word).Int.signedness == .unsigned);
+        assert(word_bits % 8 == 0); // byte multiple, so bytes can be cast to words
+        assert(@bitSizeOf(Marker) == word_bits);
+
+        assert(@bitSizeOf(MarkerRunLength) % 2 == 0);
+        assert(std.math.maxInt(MarkerRunLength) == marker_run_length_max);
+
+        assert(@bitSizeOf(MarkerLiterals) % 2 == 1);
+        assert(std.math.maxInt(MarkerLiterals) == marker_literals_max);
+    }
+
+    return struct {
+        const Self = @This();
+
+        inline fn marker(mark: Marker) Word {
+            return @bitCast(Word, mark);
+        }
+
+        /// Decodes the compressed bitset in `source` into `set`. Panics if `source`'s encoding is invalid.
+        /// Returns the number of *words* written to `target`.
+        pub fn decode(source: []const u8, target: []Word) usize {
+            const source_words = @alignCast(@alignOf(Word), mem.bytesAsSlice(Word, source));
+            var source_index: usize = 0;
+            var target_index: usize = 0;
+            while (source_index < source_words.len) {
+                const source_marker = @bitCast(Marker, source_words[source_index]);
```

Comment:
If we drop the `marker()` helper then we can also just name this as `marker` here which would thin the code out nicely.

Reply:
:+1: I renamed `marker` to `marker_word` -- it is a useful helper (especially in the tests).